content/blog/2019/10/2019-10-03-jcasc-phase1-VSCode-Case-Study.adoc

```diff
@@ -0,0 +1,92 @@
+The main premise of JCasC is to load YAML files written by developers into the Jenkins instance.An example of a jenkins YAML file is:
+
+[source, yaml]
+---
+Jenkins:
+ systemMessage: “Hello World”
+ numExecutors:2
+ ---
+
+The above YAML configuration will configure Jenkins to display a message Hello world with the number of executors set to two.In order to validate the schema we have a schema. This schema is written using jelly files (Executable XML files) and currently it is not a valid schema.The first phase of the project is based around rewriting the schema generation in java and developing a better test framework for it, because currently the schema is not testable.
+
+=== Phase 1 — JCasC Dev Tools
+The first week I got into studying how the schema was generated.With the support of two of my awesome mentors Tim Jacomb and Joseph Peterson
+I finally got an understanding of the current schema.
+So JCasC has a set of configurators for describing a YAML file.
+They are:
+a) Base Configurators
+b) Hetero Describable Configurators
+c) Data Bound Configurators
+These configurators together successfully describe a YAML file.
+We proceeded to generate the schema with the help of individual description of each of these configurators.
+The JSON Schema has a set of components, consider the above yaml file as an example:
+
+[source,json]
+---
+jenkins”: { 
+“type”: “object”, 
+“properties”: { 
+“systemMessage”: {
+ “type”: “string"},
+ “numExecutors”: {
+ “type”: “integer” 
+},
+---
+
+So here Jenkins is the base configurator and it has a set of attributes viz systemMessage and numExecutors, so our schema needs to be able to describe a set of attributes for every field in the schema. Some of the fields that our JSON Schema uses to describe the YAML are:
+
+1) type : String, int, Boolean etc.
+
+2) properties : A set of fields describing the part field.
+
+3) id: Unique Identifier for the field
+Thus the above schema successfully verifies the YAML configuration.
+
+=== JAVA Rewrite
+We used JSON Objects to build components of the schema.The basic flow that is followed to generate the schema is as follows:
+
+a) Iterate through the Base Configurators.
+
+b) Iterate over the list of Base Configurator.Attributes and add each attribute to the schema.
+
+c) Iterate over the HeteroDescribable Configurators and add each configurator to the schema along with its required properties.
+
+The set of PR’s Resolved during Phase 1 are as follows:
+
+a)  link:https://github.com/jenkinsci/configuration-as-code-plugin/pull/980[ Basic Schema]
+
+b)link:https://github.com/jenkinsci/configuration-as-code-plugin/pull/1030[ Schema Doc Page]
+
+c)link:https://github.com/jenkinsci/configuration-as-code-plugin/pull/1078[ Test Framework]
+
+d)link:https://github.com/jenkinsci/configuration-as-code-plugin/pull/1027[ Nested Schema]
+
+
+That is all from me guys, I am currently preparing for phase 2 and working towards fixing any pending issues of Phase 1. Thanks for reading.
```

Comment:
should we mention the phase two goal here?

Reply:
Yeah definitely : )